llvm/lib/Target/DirectX/DXILOpLowering.cpp

```diff
@@ -477,6 +500,11 @@ class OpLowerer {
     HasErrors |= replaceFunctionWithOp(F, OpCode);                             \
     break;
 #include "DXILOperation.inc"
+      case Intrinsic::dx_group_memory_barrier_with_group_sync:
```

Comment:
I think we could also approach this as an intrinsic expansion. So we would create a DirectX intrinsic `dx_group_memory_barrier` that would lower here. Then we would deal with `dx_group_memory_barrier_with_group_sync` and all the other barrier ops as an intrinsic expansion with the applicable opcode and barrier modes.

I think that would be clearer when adding future barrier ops and match the flow of other intrinsics. WDYT?

Reply:
Yeah make sense. I wasn't aware of the precedence of expanding one intrinsic to another, so it maps directly to a dx op.

Reply:
Here is an implementation where we can do (2) with just tablegen:

https://github.com/llvm/llvm-project/pull/112641